pageserver/src/tenant/mgr.rs

```diff
@@ -1101,101 +1313,313 @@ pub(crate) async fn attach_tenant(
+    fn upsert(mut self, new_value: TenantSlot) -> Result<(), TenantSlotUpsertError> {
+        let mut locked = TENANTS.write().unwrap();
+
+        if let TenantSlot::InProgress(_) = new_value {
+            // It is never expected to try and upsert InProgress via this path: it should
+            // only be written via the tenant_map_acquire_slot path.  If we hit this it's a bug.
+            return Err(TenantSlotUpsertError::InternalError(
+                "Attempt to upsert an InProgress state".into(),
+            ));
+        }
+
+        let m = match &mut *locked {
+            TenantsMap::Initializing => return Err(TenantMapError::StillInitializing.into()),
+            TenantsMap::ShuttingDown(_) => {
+                return Err(TenantMapError::ShuttingDown.into());
             }
-            Err(e) => Err(TenantMapInsertError::Other(e)),
-        },
+            TenantsMap::Open(m) => m,
+        };
+
+        let replaced = m.insert(self.tenant_id, new_value);
+        self.upserted = true;
+
+        // Sanity check: on an upsert we should always be replacing an InProgress marker
+        match replaced {
+            Some(TenantSlot::InProgress(_)) => {
+                // Expected case: we find our InProgress in the map: nothing should have
+                // replaced it because the code that acquires slots will not grant another
+                // one for the same TenantId.
+                Ok(())
+            }
+            None => {
+                error!(
+                    tenant_id = %self.tenant_id,
+                    "Missing InProgress marker during tenant upsert, this is a bug."
+                );
+                Err(TenantSlotUpsertError::InternalError(
+                    "Missing InProgress marker during tenant upsert".into(),
+                ))
+            }
+            Some(slot) => {
+                error!(tenant_id=%self.tenant_id, "Unexpected contents of TenantSlot during upsert, this is a bug.  Contents: {:?}", slot);
+                Err(TenantSlotUpsertError::InternalError(
+                    "Unexpected contents of TenantSlot".into(),
+                ))
+            }
+        }
     }
 }
 
-async fn tenant_map_upsert_slot<'a, F, R>(
-    tenant_id: TenantId,
-    upsert_fn: F,
-) -> Result<(), TenantMapInsertError>
-where
-    F: FnOnce(Option<TenantSlot>) -> R,
-    R: std::future::Future<Output = anyhow::Result<TenantSlot>>,
-{
-    let mut guard = TENANTS.write().await;
-    let m = match &mut *guard {
-        TenantsMap::Initializing => return Err(TenantMapInsertError::StillInitializing),
-        TenantsMap::ShuttingDown(_) => return Err(TenantMapInsertError::ShuttingDown),
+impl Drop for SlotGuard {
+    fn drop(&mut self) {
+        if !self.upserted {
+            let mut locked = TENANTS.write().unwrap();
+
+            let m = match &mut *locked {
+                TenantsMap::Initializing | TenantsMap::ShuttingDown(_) => {
+                    return;
+                }
```

Comment:
Why early-exit for `ShuttingDown` here? The `SlotGuard` could have been created earlier while we weren't shutting down.

Reply:
The ShuttingDown state never contains InProgress states -- shutdown_all_tenants0 doesn't carry them across.

One could equally copy them across during shutdown and then remove this check, but conceptually it's simpler if we think of the ShuttingDown state as being a readonly view of whatever was present when we started shutting down, rather than being mutable in the same way as the Open state.  Long term, as with other areas, I think we can live without this Open/Initializing/ShuttingDown distinction, and just not let any external APIs hit this object until we've loaded it.

Reply:
> The ShuttingDown state never contains InProgress states -- shutdown_all_tenants0 doesn't carry them across.

Right, had missed that.

I'd prefer if you could make a separate `match` branch for the `TenantsMap::ShuttingDown` case and explain that in a comment on that branch.

Reply:
Done.